Add persistent partition to RPi Pico Flash

Snap-A opened the pull request:
### Updated Summary ###
The added partition ```storage_partition``` can be used for the littelfs sample application. It is sized to have enough capacity for the created files in this sample code.

### Details ###
The changed file ```rpi_pico-common.dtsi``` now contains a third partition, that describes "storage" are for a file system. The "code" partition was reduced in size to make room for this area. The new partition has a capacity of 16 kB, which is enough for the littelfs sample to store two files.

### Tests ###

#### Compilation ####
The compile step for each platform succeeds:
```
$ west build --pristine -b rpi_pico samples/subsys/fs/littlefs
...
[160/160] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
      BOOT_FLASH:         256 B        256 B    100.00%
           FLASH:       52616 B    2080512 B      2.53%
             RAM:       11072 B       264 KB      4.10%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/awolf/zephyrproject-fork/zephyr/build/zephyr/zephyr.elf for board: rpi_pico
Converted to uf2, output size: 105984, start address: 0x10000000
Wrote 105984 bytes to zephyr.uf2
```
```
$ west build --pristine -b rpi_pico/rp2040/w samples/subsys/fs/littlefs
...
[164/164] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
      BOOT_FLASH:         256 B        256 B    100.00%
           FLASH:       52756 B    2080512 B      2.54%
             RAM:       11072 B       264 KB      4.10%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/awolf/zephyrproject-fork/zephyr/build/zephyr/zephyr.elf for board: rpi_pico
Converted to uf2, output size: 106496, start address: 0x10000000
Wrote 106496 bytes to zephyr.uf2
```

#### Test Run ####
The test were run in a Pico-W device.

##### First Run On Device #####
Log after flashing the littlefs sample:
```
Sample program to r/w files on littlefs
Area 2 at 0x1fc000 on flash-controller@18000000 for 16384 bytes
I: LittleFS version 2.10, disk version 2.1
I: FS at flash-controller@18000000:0x1fc000 is 4 0x1000-byte blocks with 512 cycle
I: partition sizes: rd 16 ; pr 16 ; ca 64 ; la 32
E: WEST_TOPDIR/modules/fs/littlefs/lfs.c:4604: Invalid block count (1 != 4)
W: can't mount (LFS -22); formatting
I: /lfs mounted
/lfs mount: 0
/lfs: bsize = 16 ; frsize = 4096 ; blocks = 4 ; bfree = 2

Listing dir /lfs ...
/lfs/boot_count read count:0 (bytes: 0)
/lfs/boot_count write new boot count 1: [wr:1]
I: Test file: /lfs/pattern.bin not found, create one!
------ FILE: /lfs/pattern.bin ------
01 55 55 55 55 55 55 55 02 55 55 55 55 55 55 55
...
45 55 aa 
I: /lfs unmounted
/lfs unmount: 0
```

After performing a power cycle:

```
Sample program to r/w files on littlefs
Area 2 at 0x1fc000 on flash-controller@18000000 for 16384 bytes
I: LittleFS version 2.10, disk version 2.1
I: FS at flash-controller@18000000:0x1fc000 is 4 0x1000-byte blocks with 512 cycle
I: partition sizes: rd 16 ; pr 16 ; ca 64 ; la 32
/lfs mount: 0
/lfs: bsize = 16 ; frsize = 4096 ; blocks = 4 ; bfree = 1

Listing dir /lfs ...
[FILE] boot_count (size = 1)
[FILE] pattern.bin (size = 547)
/lfs/boot_count read count:1 (bytes: 1)
/lfs/boot_count write new boot count 2: [wr:1]
------ FILE: /lfs/pattern.bin ------
02 55 55 55 55 55 55 55 03 55 55 55 55 55 55 55
...
46 55 ab 
I: /lfs unmounted
/lfs unmount: 0
```


## Old Text ##
The below text is kept to document previous changes to make the littlefs sample work.

**IGNORE TEXT BELOW** 
### Summary ###
The added overlay files support the compilation for the Pico and Pico_w of the sample for the littlefs filesystem.

### Details ###
The file ```rpi_pico.overlay``` reduces the size of the "code" partition and adds another partition at the end.
This partition has a capacity of 64 kB.

### Tests ###
#### Compilation ####
The compile step for each platform succeeds:
```
$ west build -b rpi_pico samples/subsys/fs/littlefs
...
-- Found BOARD.dts: /home/awolf/zephyrproject-fork/zephyr/boards/raspberrypi/rpi_pico/rpi_pico.dts
-- Found devicetree overlay: /home/awolf/zephyrproject-fork/zephyr/samples/subsys/fs/littlefs/boards/rpi_pico.overlay
-- Generated zephyr.dts: /home/awolf/zephyrproject-fork/zephyr/build/zephyr/zephyr.dts
...
Converted to uf2, output size: 105984, start address: 0x10000000
Wrote 105984 bytes to zephyr.uf2
```

```
$ west build -b rpi_pico/rp2040/w samples/subsys/fs/littlefs
...
-- Found BOARD.dts: /home/awolf/zephyrproject-fork/zephyr/boards/raspberrypi/rpi_pico/rpi_pico_rp2040_w.dts
-- Found devicetree overlay: /home/awolf/zephyrproject-fork/zephyr/samples/subsys/fs/littlefs/boards/rpi_pico_rp2040_w.overlay
-- Generated zephyr.dts: /home/awolf/zephyrproject-fork/zephyr/build/zephyr/zephyr.dts
...
Converted to uf2, output size: 106496, start address: 0x10000000
Wrote 106496 bytes to zephyr.uf2
```

#### Test Run ####
The code was run on the Pico_w test platform

1. First run included a flash erase (per config)
```
*** Booting Zephyr OS build v4.1.0-1562-g478212efc4e1 ***
Sample program to r/w files on littlefs
Area 2 at 0x1f0000 on flash-controller@18000000 for 65536 bytes
E: Erasing flash area ... 0
I: LittleFS version 2.10, disk version 2.1
I: FS at flash-controller@18000000:0x1f0000 is 16 0x1000-byte blocks with 512 cycle
I: partition sizes: rd 16 ; pr 16 ; ca 64 ; la 32
E: WEST_TOPDIR/modules/fs/littlefs/lfs.c:1389: Corrupted dir pair at {0x0, 0x1}
W: can't mount (LFS -84); formatting
I: /lfs mounted
/lfs mount: 0
/lfs: bsize = 16 ; frsize = 4096 ; blocks = 16 ; bfree = 14

Listing dir /lfs ...
/lfs/boot_count read count:0 (bytes: 0)
/lfs/boot_count write new boot count 1: [wr:1]
I: Test file: /lfs/pattern.bin not found, create one!
```
2. The second run read the per-existing files:
```
*** Booting Zephyr OS build v4.1.0-1562-g478212efc4e1 ***
Sample program to r/w files on littlefs
Area 2 at 0x1f0000 on flash-controller@18000000 for 65536 bytes
I: LittleFS version 2.10, disk version 2.1
I: FS at flash-controller@18000000:0x1f0000 is 16 0x1000-byte blocks with 512 cycle
I: partition sizes: rd 16 ; pr 16 ; ca 64 ; la 32
/lfs mount: 0
/lfs: bsize = 16 ; frsize = 4096 ; blocks = 16 ; bfree = 13

Listing dir /lfs ...
[FILE] boot_count (size = 1)
[FILE] pattern.bin (size = 547)
/lfs/boot_count read count:1 (bytes: 1)
/lfs/boot_count write new boot count 2: [wr:1]
```

Snap-A committed:
boards : rpi_pico : Add persistent partition for Pico and Pico_w

Add a 16 KB 'storage' partition at the end of the Pico flash.
That size is sufficient to hold the two test files created by
the 'littlefs' sample.

The sample 'subsys/USB/mass' uses an overlay to add the same
partition and it was changed to delete the one defined by this
DTSI file before doing that.

Signed-off-by: Andreas Wolf <awolf002@gmail.com>

ajf58 commented on samples/subsys/fs/littlefs/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,21 @@
+
+        code_partition: partition@100 {
+            label = "code-partition";
+            reg = <0x100 (DT_SIZE_M(2)-0x100-DT_SIZE_K(64))>;
```
```suggestion
            reg = <0x100 (DT_SIZE_M(2) - 0x100 - DT_SIZE_K(64))>;
```
Is there a reason not to use single whitespace either side of binary operators here?

Snap-A replied:
I will add the whitespaces. Did not even notice I "compacted" the text like that.

ajf58 commented on samples/subsys/fs/littlefs/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,21 @@
+
+        storage_partition: storage_partition@1f0000 {
+            label = "storage-partition";
+            reg = <(DT_SIZE_M(2)-DT_SIZE_K(64)) (DT_SIZE_K(64))>;
```
```suggestion
            reg = <(DT_SIZE_M(2) - DT_SIZE_K(64)) (DT_SIZE_K(64))>;
```
Same comment as above. Is there a reason not to use single whitespace either side of binary operators here?

Snap-A replied:
Same here.

Regarding the merge commit, I had re-merged the branch before the PR creation. Did not realize this would be visible, here.

In the Contribution Guidelines, it says to use ```--rebase <commit>^```, and then a ```push --force``` to update this PR. I will try this out, it may clean up things...

Additionally, I missed the "copyright / license" comment lines at the start of these files. Will fix that, too.

ajf58 reviewed (commented):
Functionally this looks good. It looks like you've got a merge commit that probably doesn't want to be here, however.

Snap-A commented on samples/subsys/fs/littlefs/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,26 @@
+/delete-node/ &code_partition;
+
+&flash0 {
+    partitions {
```
Will fix the tabs.
Not sure if the change is welcomed in the board file, but I'm open to adjust the PR to place it in there if people agree.

ajf58 replied:
> Not sure if the change is welcomed in the board file,

I think it's a good idea.

I think my questions would be (sharing my train of thought):
1. What's the idiomatic thing in the Zephyr codebase? A very crude search suggests, yes, a lot of boards have a storage partition defined for a good "out-of-the-box" experience.
2. What is motivating the choice of size? > 1 flash sector seems like a minimum. Has something motivated the choice of 64kB in particular?

Snap-A replied:
Regarding (2):
The choice of 64 kB gave the FS 16 blocks to work with. That seemed sufficient for the root dir and another few files (15 if each is smaller than a couple of kB in size).

Space for only two files are needed in this specific example and so it could be reduced to 16 kB (4 blocks), but the Pico has 2MB flash, so the 64 kB seem seemed to be more realistic to me.

ajf58 commented on samples/subsys/fs/littlefs/boards/rpi_pico.overlay:
```diff
@@ -0,0 +1,26 @@
+            read-only;
+        };
+
+        storage_partition: storage_partition@1f0000 {
+            label = "storage-partition";
```
```suggestion
        storage_partition: partition@1f0000 {
            label = "storage";
```
Seems to be a more common idiom in the codebase: I think partitions labelled `<foo>-partition` is perhaps a bit redundant.

Snap-A replied:
That name could be addressed if this is made a change to the Pico board dts. As an overlay, it probably should stick with the base naming scheme...

ajf58 commented on samples/subsys/fs/littlefs/boards/rpi_pico_rp2040_w.overlay:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2025 Andreas Wolf <awolf002@gmai.com>
```
```suggestion
 * Copyright (c) 2025 Andreas Wolf <awolf002@gmail.com>
```

Snap-A replied:
Fixing that.

Snap-A commented:
* Reduced the size of the partition to a 2^n number that is large enough to hold the two test files;
* Fixed the line where I missed the white space change (' ' -> '\t') 
* Adjusted to commit message;

Snap-A commented:
I adjusted the PR to a change of the DTS include file ```boards/raspberrypi/rpi_pico/rpi_pico-common.dtsi``` and removed the added overlay files.

The plan is to document in the description the tests that were run with this change, next.

Please give feedback if this would be acceptable as a PR.

Snap-A commented:
The twister build found a problem in the sample code for USB mass storage: It also defines a "storage_partition". This will fail with this proposed change, unless it is "deleted" and re-defined in the overlay file.

I will add this change to this PR, so that the twister is "happy".

soburi commented on boards/raspberrypi/rpi_pico/rpi_pico-common.dtsi:
```diff
@@ -79,13 +79,24 @@
+		 * of the code partition. The size of 16 KB is chosen to have the capacity for
+		 * a minimal 'littlefs' file system with 4 blocks.
+		 */
+		storage_partition: partition@1fc000 {
+			label = "storage-partition";
+			reg = <(DT_SIZE_M(2) - DT_SIZE_K(16)) (DT_SIZE_K(16))>;
```
This is fine, but the @xxxxxxxx at the end of the node path and the value at the beginning of reg must match. So, represent in the calculations make bit hard to understand.

IMO, If you want to make it easier to read, you might want to introduce

#define STORAGE_PARTITION_SIZE DT_SIZE_K(16)

Snap-A replied:
I will take a look at that. This should improved readability, I agree.

de-nordic commented:
Re-assigning to RPI maintainer.

ajf58 commented on boards/raspberrypi/rpi_pico/rpi_pico-common.dtsi:
```diff
@@ -62,8 +62,11 @@
 	};
 };
 
+#define FLASH_SIZE		DT_SIZE_M(2)
+#define STORAGE_PARTITION_SIZE	DT_SIZE_K(16)
```
Having a  `#define`s in the middle of a .dtsi file doesn't seem very idiomatic. Have you come across this elsewhere in the codebase?

I appreciate what you're trying to do here, however, but I think for reasons I don't understand, this is frowned upon in the codebase.

soburi reviewed (commented):
https://github.com/zephyrproject-rtos/zephyr/pull/90411
This review seems to have been taken first.

https://github.com/zephyrproject-rtos/zephyr/pull/90758
If this change is also included,

will the `rpi_pico-common.dtsi` fix already be included?

If so, only the overlay part is needed, so please fix the patch.

soburi commented:
@Snap-A 
Yes. If no additional modifications are required, I think it's best to close it.

Snap-A commented:
I tested compiling the USB "mass storage" sample with the changes in #90758 and no further changes to the sample code.
This succeeded, and the created binary seems to successfully run on the Pico, since I see the mass storage device in the USB list:
```
Bus 002 Device 007: ID 2fe3:0008 NordicSemiconductor Zephyr MSC sample
```
I am closing this PR.
